File: Java/MergeSort.java
import java.io.*;
public class MergeSort 
{
 
   public static void merge(int a[],int l,int m,int h)
   {  
       int i, j,c=l;
       int b[]=new int[h+1];
 
	for(i = l,j = m+1; i<=m && j<=h; c++)
             {
 
                  if(a[i] <= a[j])
                  b[c] = a[i++];
                 else
                 b[c] = a[j++];
      	}
	 while(i <= m )    
                 b[c++] = a[i++];
 
               while(j<=h)   
                 b[c++] = a[j++];
          
   	 for(i = l ; i <= h; i++)
                        a[i] = b[i];
   } 
  
   public static void Sort(int a[],int l,int h)
   {  
         if(l<h)
        {
           int m=(l+h)/2;
           Sort(a,l,m);
           Sort(a,m+1,h);
           merge(a,l,m,h);
 
         } 
          
   
   }  
 
 public static void printarray(int a[])
{
       for(int i=0; i < a.length; i++)
        {
        
        System.out.print(a[i]+"  ");
        }
       
}
public static void main(String[] args) throws IOException
{
 
    int n,i;
     BufferedReader b=new BufferedReader(new InputStreamReader(System.in));
 
  
   System.out.println("enter N: ");
 n=Integer.parseInt(b.readLine());
int a[] = new int[n];  
  System.out.println("enter "+n+" elements ");
 for(i= 0; i< n; i++)
 
  a[i] = Integer.parseInt(b.readLine()); 
 
  System.out.println("elements in array ");
        printarray(a);
        Sort(a,0,n-1);
         System.out.println("\nelements after sorting");
         printarray(a);
       
  }
}

File: C#/MergeSort.cs
using System;
using System.Collections.Generic;
using System.Linq;

class MergeSort
{
    static void Main(string[] args)
    {
        // Based on the algorithm found here
        // https://www.tutorialspoint.com/data_structures_algorithms/merge_sort_algorithm.htm
        IList<int> elements = new List<int>();
        Console.WriteLine($"Enter array members:");
        int i = 1;
        while (true)
        {
            Console.Write($"Enter integer element {i}, anything else to sort:");
            string unparsedElement = Console.ReadLine();
            bool parsePass = int.TryParse(unparsedElement, out int element);
            if (parsePass)
            {
                elements.Add(element);
                i++;
            }
            else
            {
                break;
            }
        }

        Console.WriteLine("elements in array ");
        Printarray(elements);
        IList<int> sortedElements = Sort(elements);
        Console.WriteLine("elements after sorting");
        Printarray(sortedElements);

    }

    private static IList<int> Sort(IList<int> elements)
    {
        if (elements.Count == 1)
        {
            return elements;
        }

        IList<int> elementsLeftHalf = elements.Take(elements.Count / 2).ToList();
        IList<int> elementsRightHalf = elements.Skip(elements.Count / 2).ToList();
        IList<int> sortedLeftHalf = Sort(elementsLeftHalf);
        IList<int> sortedRightHalf = Sort(elementsRightHalf);
        return Merge(sortedLeftHalf, sortedRightHalf);
    }

    private static IList<int> Merge(IList<int> sortedLeftHalf, IList<int> sortedRightHalf)
    {
        IList<int> mergedSortedElements = new List<int>();
        while (sortedLeftHalf.Any() && sortedRightHalf.Any())
        {
            if (sortedLeftHalf[0] > sortedRightHalf[0])
            {
                mergedSortedElements.Add(sortedRightHalf[0]);
                sortedRightHalf.RemoveAt(0);
            }
            else
            {
                mergedSortedElements.Add(sortedLeftHalf[0]);
                sortedLeftHalf.RemoveAt(0);
            }
        }

        while (sortedLeftHalf.Any())
        {
            mergedSortedElements.Add(sortedLeftHalf[0]);
            sortedLeftHalf.RemoveAt(0);
        }

        while (sortedRightHalf.Any())
        {
            mergedSortedElements.Add(sortedRightHalf[0]);
            sortedRightHalf.RemoveAt(0);
        }

        return mergedSortedElements;

    }

    private static void Printarray(IList<int> elements)
    {
        foreach (var element in elements)
        {
            Console.Write($"{element},");
        }
        Console.WriteLine();
    }
}
